runtime/compiler/env/PersistentAllocator.cpp

```diff
@@ -228,58 +324,256 @@ void
+                  // which case _endInterval needs to be changed to block
+                  if (_endInterval[index]->size() < block->size())
+                     {
+                     TR_ASSERT(getInterval(blockIterator->size()) > index, "blockInterator must be first block in the next interval: index=%zu b->sz=%zu bi->sz=%zu", index, block->size(), blockIterator->size());
+                     _endInterval[index] = block;
+                     }
+                  }
+               }
+            }
+         else // Insert after 'prev'; block will be the very last entry in the list
+            {
+            TR_ASSERT(prev, "prev must exist");
+            TR_ASSERT(!prev->next(), "this must be the last element in the list");
+            block->setPrevious(prev);
+            block->setNext(NULL);
+            prev->setNext(block);
+            _endInterval[index] = block;
+            }
+         }
+      else // There are no other blocks in this interval
+         {
+         // Find the closest interval that has a block
+         // I verified that the block will not be attached at the beginning of the list.
+         // Thus, there must be a previous interval with some block in it
+         TR_ASSERT(index > 0, "Index must be greater than 0");
+         for (int i = (int)index - 1; i >= 0; i--)
+            {
+            if (_endInterval[i])
+               {
+               block->setPrevious(_endInterval[i]);
+               block->setNext(_endInterval[i]->next());
+               TR_ASSERT(block->size() > _endInterval[i]->size(), "wrong sizes previous");
+               if (block->next())
+                  {
+                  block->next()->setPrevious(block);
+                  TR_ASSERT(block->size() < block->next()->size(), "wrong sizes after block->size()=%zu next->size()=%zu", block->size(), block->next()->size());
+                  }
+               _endInterval[i]->setNext(block);
+               break;
+               }
+            }
+         TR_ASSERT(block->previous(), "I must have attached my block to the list");
+         _startInterval[index] = block;
+         _endInterval[index] = block;
+         }
       }
+      checkIntegrity("freeVariableSizeBlock end");   
    }
 
+void 
+PersistentAllocator::checkIntegrity(const char msg[])
+   {
+#ifdef DEBUG
+   ExtendedBlock *firstBlock = reinterpret_cast<ExtendedBlock*>(_freeBlocks[LARGE_BLOCK_LIST_INDEX]);
+   ExtendedBlock *prevBlock = NULL;
+   TR_ASSERT_FATAL(!firstBlock || firstBlock->previous() == NULL, "Error for first block");
+   for (ExtendedBlock *block = firstBlock; block; prevBlock = block, block = block->next())
+      {
+      if (prevBlock)
+         {
+         TR_ASSERT_FATAL(prevBlock->size() < block->size(), "Blocks are not correctly ordered");
+         TR_ASSERT_FATAL(block->previous() == prevBlock, "block->previous() != prevBlock");
+         TR_ASSERT_FATAL(prevBlock->next() == block, "prevBlock->next() != block");
```

Comment:
I don't think this is necessary, since `prevBlock` is set to `block`, and `block` is set to `block->next()`, so `prevBlock->next()` is always going to be `block`, except for I guess maybe some race condition.

Reply:
Correct. I'll delete the last assert.